### PR TITLE
2862 using paketo buildpack for liberty on azure and google cloud

### PR DESCRIPTION
**Pull request by kevin-ortega**
@GraceJansen 

Link to [blog](https://blogs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/blog/2023/02/17/using-liberty-buildpack-on-azure-and-google-cloud.html) on draft site.  

Open issue, #2862 

image
image
image

### CODE REVIEW

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+= Deploy Java applications to Azure and Google Cloud using Paketo Buildpacks
+Kevin Ortega <https://github.com/kevin-ortega>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+
+Azure and Google Cloud have built-in support for link:https://buildpacks.io[cloud native buildpacks].  
+You can use link:https://learn.microsoft.com/en-us/azure/container-registry/container-registry-tasks-pack-build[Azure] or link:https://cloud.google.com/docs/buildpacks/build-application[Google Cloud] to easily deploy Java applications by using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your source code or container images.
+
+This post provides instructions on how to use the Paketo Buildpack for Liberty with Azure and Google Cloud. To learn more about using Azure and Google Cloud see their respective documentation.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
This post provides instructions on how to use the Paketo Buildpack for Liberty with Azure and Google Cloud. To learn more about using Azure and Google Cloud, see their respective documentation.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+:url-about: /
+
+Azure and Google Cloud have built-in support for link:https://buildpacks.io[cloud native buildpacks].  
+You can use link:https://learn.microsoft.com/en-us/azure/container-registry/container-registry-tasks-pack-build[Azure] or link:https://cloud.google.com/docs/buildpacks/build-application[Google Cloud] to easily deploy Java applications by using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your source code or container images.
+
+This post provides instructions on how to use the Paketo Buildpack for Liberty with Azure and Google Cloud. To learn more about using Azure and Google Cloud see their respective documentation.  
+
+== What do you need?
+* An https://azure.microsoft.com/en-us/free/[Azure] or https://cloud.google.com/free/[Google Cloud] Account
+* The link:https://learn.microsoft.com/en-us/cli/azure/install-azure-cli[Azure cli] or
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
* The link:https://learn.microsoft.com/en-us/cli/azure/install-azure-cli[Azure cli] or the link:https://cloud.google.com/sdk/docs/install[gcloud cli]
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+* The link:https://cloud.google.com/sdk/docs/install[gcloud cli]
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+* Your application source
+
+To build and deploy container images or create a link:https://buildpacks.io/docs/concepts/components/builder/[custom builder], you'll also need:
+
+* The cloud native buildpack link:https://buildpacks.io/docs/tools/pack/[pack CLI]
+* https://www.docker.com/[Docker] or https://podman.io[Podman]. This post focuses on Docker. For information about using Podman with Paketo, refer to the buildpacks.io documentation for https://buildpacks.io/docs/app-developer-guide/building-on-podman/[Building on Podman]. 
+* An externally accessible container registry that can host your images.
+
+The Paketo Buildpack for Liberty is included in the composite Paketo Buildpack for Java that also includes `tomcat` and `tomee` application servers. Tomcat is the default application server and to use Open Liberty requires a little extra configuration.  The Paketo Buildpack for Java includes the Bellsofit Liberica as the default JVM provider.  To use other JVM providers requires extra configuration.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
The Paketo Buildpack for Liberty is included in the composite Paketo Buildpack for Java that also includes `tomcat` and `tomee` application servers. Tomcat is the default application server and to use Open Liberty requires a little extra configuration.  The Paketo Buildpack for Java includes Bellsoft Liberica as the default JVM provider.  To use other JVM providers requires extra configuration.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+You can set the application server that the Paketo Buildpack for Java uses by specifying the `BP_JAVA_APP_SERVER` environment variable and select an alternate JVM provider using a link:https://buildpacks.io/docs/app-developer-guide/using-project-descriptor/[project descriptor]. 
+
+== Let's get started
+
+Create the project descriptior file, `project.toml`, in the application's root source directory. At a minimum, `project.toml` should contain `BP_JAVA_APP_SERVER=liberty` but can contain other environment variables consumed by buildpacks.  To use the alternate and preferred JVM, eclipse openj9, we change the order that the buildpacks run in the project descriptor.   To minimize the size of the application container image, the buildpack by default installs the kernel profile.  You can override the profile using the `BP_LIBERTY_PROFILE` environment variable or use the `BP_LIBERTY_FEATURES` environment variable to install the features required by the application.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
Create the project descriptior file, `project.toml`, in the application's root source directory. At a minimum, `project.toml` must  specify the `BP_JAVA_APP_SERVER=liberty` environment variable but can contain other environment variables that are consumed by buildpacks.  

To use the alternate and preferred Eclipse OpenJ9 JVM,  change the order that the buildpacks run in the project descriptor.  To minimize the size of the application container image, the buildpack installs the Liberty kernel profile by default.  You can override the profile using the `BP_LIBERTY_PROFILE` environment variable or use the `BP_LIBERTY_FEATURES` environment variable to install the Liberty features your application requires.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+You can set the application server that the Paketo Buildpack for Java uses by specifying the `BP_JAVA_APP_SERVER` environment variable and select an alternate JVM provider using a link:https://buildpacks.io/docs/app-developer-guide/using-project-descriptor/[project descriptor]. 
+
+== Let's get started
+
+Create the project descriptior file, `project.toml`, in the application's root source directory. At a minimum, `project.toml` should contain `BP_JAVA_APP_SERVER=liberty` but can contain other environment variables consumed by buildpacks.  To use the alternate and preferred JVM, eclipse openj9, we change the order that the buildpacks run in the project descriptor.   To minimize the size of the application container image, the buildpack by default installs the kernel profile.  You can override the profile using the `BP_LIBERTY_PROFILE` environment variable or use the `BP_LIBERTY_FEATURES` environment variable to install the features required by the application.  
+
+For example, your project descriptor should contain the following content to use the liberty applicaiton server, install eclipse openj9, install the kernel profile and install a set of liberty features:
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
For example, your project descriptor might contain the following content to use the Liberty runtime and install Eclipse OpenJ9, the Liberty kernel profile, and a set of Liberty features:
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+[[build.buildpacks]]
+  uri = "docker://gcr.io/paketo-buildpacks/eclipse-openj9"
+  
+[[build.buildpacks]]
+  uri = "docker://gcr.io/paketo-buildpacks/java"
+```
+
+== Using Azure
+Use the link:https://learn.microsoft.com/en-us/cli/azure/acr/pack?view=azure-cli-latest#az_acr_pack_build[az acr pack build] command to build and push a container image using the built-in cloud native buildpack support.   
+For example, the following command builds and pushes the container image using the `mycr` Azure container registry and the full paketo buildacks builder.  The resulting image name is `myapp` and the application source resides in `~/myappsourcedir`.  The application source could also be the url of a git repo.  This example assumes the `mycr` Azure container registry was previously created.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
For example, the following command builds and pushes the container image using the `mycr` Azure container registry and the full Paketo Buildpacks builder.  The resulting image name is `myapp` and the application source resides in `~/myappsourcedir`.  The application source can also be specified as the URL of a Git repo.  This example assumes the `mycr` Azure container registry was previously created.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    value = "jsonb-2.0 mpconfig-3.0 mpmetrics-4.0 restfulws-3.0 jsonp-2.0 cdi-3.0"     
+    
+[[build.buildpacks]]
+  uri = "docker://gcr.io/paketo-buildpacks/eclipse-openj9"
+  
+[[build.buildpacks]]
+  uri = "docker://gcr.io/paketo-buildpacks/java"
+```
+
+== Using Azure
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
== Build and push a container image with Azure
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+```
+az acr pack build 
+    --registry mycr 
+    --pull --builder paketobuildpacks/builder:full 
+    --image myapp 
+    ~/myappsourcedir
+```
+
+== Using Google Cloud
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
== Build and push a container image with Google Cloud
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+Use the link:https://cloud.google.com/docs/buildpacks/build-application#remote_builds[gcloud builds] command to build and push a container image using the built-in cloud native buildpack support.  
+Follow the link:https://cloud.google.com/docs/buildpacks/build-application#before-you-begin[Before you begin] instructions to setup your environment.  
+
+The basic command syntax is:
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
The command uses the following basic syntax:
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+```
+gcloud builds submit --pack image=LOCATION-docker.pkg.dev/PROJECT_ID/REPO_NAME/IMAGE_NAME
+```
+
+In the following example, `us-east1` is the LOCATION, `paketobuildpacks` is the PROJECT_ID, `paketobuildpacktest`` is the REPO_NAME and `javatest` is the IMAGE_NAME.  The command is executed from the application's root directory.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
In the following example, `us-east1` is the LOCATION, `paketobuildpacks` is the PROJECT_ID, `paketobuildpacktest` is the REPO_NAME and `javatest` is the IMAGE_NAME.  The command is executed from the application's root directory.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    image=us-east1-docker.pkg.dev/paketobuildpacks/paketobuildpacktest/javatest,builder=paketobuildpacks/builder:full
+
+```
+
+== Other environment variables you can set in the `project.toml` file
+BP_LIBERTY_INSTALL_TYPE::
+Specifies the link:https://github.com/paketo-buildpacks/liberty#install-types[Install type] of Liberty. Open Liberty (value of `ol`) is the default.  Use `wlp` for WebSphere Liberty.  
+
+BP_LIBERTY_PROFILE::
+Specifies which liberty profile to install. Valid profiles for Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[in the buildpacks documentation].  The default profile is `kernel`.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
Specifies which Liberty profile to install. Valid profiles for Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[in the buildpacks documentation]. The default profile is `kernel`.  
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+In the following example, `us-east1` is the LOCATION, `paketobuildpacks` is the PROJECT_ID, `paketobuildpacktest`` is the REPO_NAME and `javatest` is the IMAGE_NAME.  The command is executed from the application's root directory.  
+
+```
+gcloud builds submit --pack 
+    image=us-east1-docker.pkg.dev/paketobuildpacks/paketobuildpacktest/javatest,builder=paketobuildpacks/builder:full
+
+```
+
+== Other environment variables you can set in the `project.toml` file
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
== Liberty environment variables you can set in the `project.toml` file
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+BP_LIBERTY_INSTALL_TYPE::
+Specifies the link:https://github.com/paketo-buildpacks/liberty#install-types[Install type] of Liberty. Open Liberty (value of `ol`) is the default.  Use `wlp` for WebSphere Liberty.  
+
+BP_LIBERTY_PROFILE::
+Specifies which liberty profile to install. Valid profiles for Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[in the buildpacks documentation].  The default profile is `kernel`.  
+
+BP_LIBERTY_FEATURES::
+Specifies a space-separated list of Liberty features to be installed with the Liberty runtime. Supports any valid Liberty feature.
+
+== Taking full advantage of all what the Paketo Buildpack for Liberty has to offer
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
== Take full advantage of all what the Paketo Buildpack for Liberty with the pack build CLI
```

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+Some features of the Paketo Buildpack for Liberty are not easily available to Azure or Google Cloud. Features like link:https://github.com/paketo-buildpacks/liberty/blob/main/docs/installing-ifixes.md[installing iFixes], link:https://github.com/paketo-buildpacks/liberty#using-custom-features[custom features], and installing from a link:https://github.com/paketo-buildpacks/liberty#building-from-a-packaged-server[packaged server] or link:https://github.com/paketo-buildpacks/liberty#building-from-a-liberty-server[server directory] aren't available when you use the built-in support to create the container image.
+
+For these features, you can use the `pack build` CLI to create the container image, tag and push the image to an external container registry. Then, use Azure or Google Cloud to deploy and manage your container by pulling your container image from the container registry.
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]: Can we provide a link here to documentation about using pack build cli with Liberty? Maybe in the liberty paketo doc or https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review, let me know if you have any questions. Thanks

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc (hunk 15 above)
Comment (kevin-ortega) [reply posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]: Yes, good suggestion.  The link to the blog is the best one....really the only place when we have any sort of documentation on using the pack cli with Liberty.   I'll add a link.

Comment (kevin-ortega) [reply posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]: Update made.

File: posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+BP_LIBERTY_INSTALL_TYPE::
+Specifies the link:https://github.com/paketo-buildpacks/liberty#install-types[Install type] of Liberty. Open Liberty (value of `ol`) is the default.  Use `wlp` for WebSphere Liberty.  
+
+BP_LIBERTY_PROFILE::
+Specifies which Liberty profile to install. Valid profiles for Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[in the buildpacks documentation]. The default profile is `kernel`.  
+
+BP_LIBERTY_FEATURES::
+Specifies a space-separated list of Liberty features to be installed with the Liberty runtime. Supports any valid Liberty feature.
+
+== Take full advantage of all what the Paketo Buildpack for Liberty with the pack build CLI
```
Comment (dmuelle) [inline posts/2023-02-17-using-liberty-buildpack-on-azure-and-google-cloud.adoc]:
```suggestion
== Take full advantage of the Paketo Buildpack for Liberty with the pack build CLI
```

**Review by dmuelle — APPROVED** [review]
LGTM, thanks

**Pull request merged**